core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java

```diff
@@ -466,700 +626,845 @@ private void buildChannel() {
                channel = new JChannel(new ByteArrayInputStream(cfg.getBytes()));
             } catch (Exception e) {
                throw log.errorCreatingChannelFromConfigString(cfg, e);
-
             }
          }
       }
 
       if (channel == null) {
          log.unableToUseJGroupsPropertiesProvided(props);
          try {
-            channel = new JChannel(fileLookup.lookupFileLocation(DEFAULT_JGROUPS_CONFIGURATION_FILE, configuration.classLoader()));
+            channel = new JChannel(
+                  fileLookup.lookupFileLocation(DEFAULT_JGROUPS_CONFIGURATION_FILE, configuration.classLoader()));
          } catch (Exception e) {
             throw log.errorCreatingChannelFromConfigFile(DEFAULT_JGROUPS_CONFIGURATION_FILE, e);
          }
       }
    }
 
-   // ------------------------------------------------------------------------------------------------------------------
-   // querying cluster status
-   // ------------------------------------------------------------------------------------------------------------------
-
-   @Override
-   public boolean isCoordinator() {
-      return isCoordinator;
+   private void receiveClusterView(View new_view) {
+      if (address == null) {
+         address = fromJGroupsAddress(channel.getAddress());
+      }
+      List<List<Address>> subGroups;
+      if (new_view instanceof MergeView) {
+         CLUSTER.receivedMergedView(channel.clusterName(), new_view);
+         subGroups = new ArrayList<>();
+         List<View> jgroupsSubGroups = ((MergeView) new_view).getSubgroups();
+         for (View group : jgroupsSubGroups) {
+            subGroups.add(fromJGroupsAddressList(group.getMembers()));
+         }
+      } else {
+         CLUSTER.receivedClusterView(channel.clusterName(), new_view);
+         subGroups = Collections.emptyList();
+      }
+      receiveClusterView(new_view.getViewId().getId(), fromJGroupsAddressList(new_view.getMembers()), subGroups);
    }
 
-   @Override
-   public Address getCoordinator() {
-      return coordinator;
+   private static List<Address> fromJGroupsAddressList(List<org.jgroups.Address> list) {
+      if (list == null || list.isEmpty())
+         return Collections.emptyList();
+
+      List<Address> retval = new ArrayList<>(list.size());
```

Comment:
The listeners are invoked without copying the subgroups lists, so I'd rather keep it immutable.